core/src/main/scala/akka/kafka/internal/ProducerStage.scala

```diff
@@ -6,126 +6,284 @@
+    }
+    else {
+      new ProducerStageLogic(shape, _producer, _inheritedAttributes)
+    }
+  }
 
-      override protected def logSource: Class[_] = classOf[ProducerStage[K, V, P]]
+  /**
+   * Default Producer State Logic
+   */
+  class ProducerStageLogic(shape: Shape, producer: Producer[K, V], inheritedAttributes: Attributes) extends TimerGraphStageLogic(shape) with StageLogging with MessageCallback[K, V, P] with ProducerCompletionState {
+    lazy val decider = inheritedAttributes.get[SupervisionStrategy].map(_.decider).getOrElse(Supervision.stoppingDecider)
+    val awaitingConfirmation = new AtomicInteger(0)
+    @volatile var inIsClosed = false
+    var completionState: Option[Try[Unit]] = None
 
-      def checkForCompletion() = {
-        if (isClosed(in) && awaitingConfirmation.get == 0) {
-          completionState match {
-            case Some(Success(_)) => completeStage()
-            case Some(Failure(ex)) => failStage(ex)
-            case None => failStage(new IllegalStateException("Stage completed, but there is no info about status"))
-          }
+    override protected def logSource: Class[_] = classOf[ProducerStage[K, V, P]]
+
+    def checkForCompletion() = {
+      if (isClosed(in) && awaitingConfirmation.get == 0) {
+        completionState match {
+          case Some(Success(_)) =>
+            onCompletionSuccess()
+            completeStage()
+          case Some(Failure(ex)) =>
+            onCompletionFailure(ex)
+            failStage(ex)
+          case None => failStage(new IllegalStateException("Stage completed, but there is no info about status"))
         }
       }
+    }
+
+    val checkForCompletionCB = getAsyncCallback[Unit] { _ =>
+      checkForCompletion()
+    }
+
+    val failStageCb = getAsyncCallback[Throwable] { ex =>
+      failStage(ex)
+    }
 
-      val checkForCompletionCB = getAsyncCallback[Unit] { _ =>
+    override val onMessageAckCb: AsyncCallback[Message[K, V, P]] = getAsyncCallback[Message[K, V, P]] { _ => }
+
+    setHandler(out, new OutHandler {
+      override def onPull() = tryPull(in)
+    })
+
+    setHandler(in, new InHandler {
+      override def onPush() = produce(grab(in))
+
+      override def onUpstreamFinish() = {
+        inIsClosed = true
+        completionState = Some(Success(()))
         checkForCompletion()
       }
 
-      val failStageCb = getAsyncCallback[Throwable] { ex =>
-        failStage(ex)
+      override def onUpstreamFailure(ex: Throwable) = {
+        inIsClosed = true
+        completionState = Some(Failure(ex))
+        checkForCompletion()
       }
+    })
 
-      setHandler(out, new OutHandler {
-        override def onPull() = {
-          tryPull(in)
+    def produce(msg: Message[K, V, P]): Unit = {
+      val r = Promise[Result[K, V, P]]
+      producer.send(msg.record, new Callback {
+        override def onCompletion(metadata: RecordMetadata, exception: Exception) = {
```

Comment:
Agreed, but I'm not sure there's much value in doing it in this one use case where we handle a producer [`Callback`](https://kafka.apache.org/10/javadoc/org/apache/kafka/clients/producer/Callback.html).  The `produce` method's code was moved, but otherwise unchanged from its original implementation.